### PR TITLE
Remove multiple FilenameFilter

### PR DESCRIPTION
**Pull request by Beacze**
Separate  redundant override accept method for FilenameFilter interface. Move them to separated classes, which implement this interface. Use those classes.

### CODE REVIEW

File: Kitodo/src/main/java/org/kitodo/filters/FileBeginEndFilter.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+    /**
+     * Filter files by given begin and end of file.
+     *
+     * @param end
+     *            String
+     */
+    public FileBeginEndFilter(String begin, String end) {
+        if (begin == null || end == null) {
+            throw new IllegalArgumentException("No filter for file begin or end given.");
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/org/kitodo/filters/FileBeginEndFilter.java]: - Should this compare not include empty string argument and not only null?
- I'm a little bit unsure about this not declared IllegalArgumentException. Maybe this exception should descriped in the JavaDoc part.

File: Kitodo/src/main/java/org/kitodo/filters/FileDirectoryFilter.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+import java.io.FilenameFilter;
+
+/**
+ * Implementation of FileNameFilter for checking if file is a directory.
+ */
+public class FileDirectoryFilter implements FilenameFilter {
+
+    @Override
+    public boolean accept(File dir, String name) {
+        return (new File(dir, name).isDirectory());
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/org/kitodo/filters/FileDirectoryFilter.java]: One more unsure thing: Would this new File declaration not left a unused resource left after this method is closed?

File: Kitodo/src/main/java/org/kitodo/filters/FileEndFilter.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+    /**
+     * Filter files by given file end.
+     *
+     * @param end
+     *            String
+     */
+    public FileEndFilter(String end) {
+        if (end == null) {
+            throw new IllegalArgumentException("No filter for file end given.");
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/org/kitodo/filters/FileEndFilter.java]: Same questions like above.

File: Kitodo/src/main/java/org/kitodo/filters/FileEndNotBeginFilter.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+    /**
+     * Filter files by given begin and end of file.
+     *
+     * @param end
+     *            String
+     */
+    public FileEndNotBeginFilter(String notBegin, String end) {
+        if (notBegin == null || end == null) {
+            throw new IllegalArgumentException("No filter for file begin or end given.");
```
Comment (henning-gerhardt) [inline Kitodo/src/main/java/org/kitodo/filters/FileEndNotBeginFilter.java]: Same like above.

File: Kitodo/src/main/java/org/kitodo/filters/FileDirectoryFilter.java (hunk 2 above)
Comment (Beacze) [inline Kitodo/src/main/java/org/kitodo/filters/FileDirectoryFilter.java]: I don't think so.

File: Kitodo/src/main/java/org/kitodo/filters/FileBeginEndFilter.java (hunk 1 above)
Comment (Beacze) [inline Kitodo/src/main/java/org/kitodo/filters/FileBeginEndFilter.java]: - Yes, comparing to empty String is a good idea.
- I can add it do JavaDoc.

File: Kitodo/src/main/java/org/kitodo/services/file/FileService.java [hunk 5]
```diff
@@ -637,14 +638,9 @@ public boolean deleteProcessContent(Process process) {
      */
     public URI getSourceDirectory(Process process) {
         URI dir = getProcessSubTypeURI(process, ProcessSubType.IMAGE, null);
-        FilenameFilter filterVerz = new FilenameFilter() {
-            @Override
-            public boolean accept(File dir, String name) {
-                return (name.endsWith("_" + "source"));
-            }
-        };
+        FilenameFilter filterDirectory = new FileEndFilter("_source");
```
Comment (Kathrin-Huber) [inline Kitodo/src/main/java/org/kitodo/services/file/FileService.java]: the variable name does not match the FileEndFilter. It is a filter FOR the directory. if you want to reflect that, then name it filterForDirectory

**Comment by Kathrin-Huber** [comment]
I don't like the classnames
FileEndNotBeginFilter.java - I get it from the context, but the classname is pretty confusing.
same for
FileListFilter - is it filtered by a list?

FileDirectoryFilter - get all files in a directory?

You see, it's not uniquely. Maybe some verbes would help, like FileIsDirectoryFilter.

**Comment by Beacze** [comment]
FileListFilter - I suppose (it is not my class) that it filters list of the files with given name

Which names you would suggest for FileEndNotBeginFilter?

**Comment by henning-gerhardt** [comment]
There is one merge conflict. Please resolve this conflict.

**Comment by henning-gerhardt** [comment]
Some name suggestions
- `FileDirectoryFilter` should be renamed to `DirectoryFilter` as this filter only checks if a given `File` object is a directory (alternative name could be  `IsDirectoryFilter`)
- `FileListFilter` is used for filtering files inside a directory with a general match as is could be called `FilennameMatchFilter`
- `FileEndFilter` could be renamed to `FilenameEndFilter`
- `FileNotEndFilter` could be renamed to `FilenameNotEndFilter`
- `FileBeginEndFilter` could be renamed to `FilenameBeginAndEndFilter`

What did you think @Beacze and @Kathrin-Huber ?

**Comment by Kathrin-Huber** [comment]
Henning wrote: 

> Some name suggestions
> * FileDirectoryFilter should be renamed to DirectoryFilter as this filter only checks if a given File object is a directory (alternative name could be IsDirectoryFilter)
> * FileListFilter is used for filtering files inside a directory with a general match as is could be called FilennameMatchFilter
> * FileEndFilter could be renamed to FilenameEndFilter
> * FileNotEndFilter could be renamed to FilenameNotEndFilter
> * FileBeginEndFilter could be renamed to FilenameBeginAndEndFilter

I would really urge to use verbs
So my suggestions:

* IsDirectoryFilter
* MatchesFileNameFilter
* FileNameEndsWithFilter
* FileNameDoesNotEndWithFilter
* FileNameBeginsWithAndEndsWithFilter


Those are pretty stupid names and they are long... why do we need this classes? Can't we just make methods in the FileService?

**Comment by henning-gerhardt** [comment]
@Kathrin-Huber : I tried to avoid long class names but I would prefer your suggestions.

We must use filter classes as long we are using the `listFiles()` method from `Files` class. Maybe from newer `java.nio.files` package we can use other possibilities.

**Comment by henning-gerhardt** [comment]
In my opinion this PR is good now. 

@Kathrin-Huber: what did you think?

**Pull request merged**